File: library/src/main/java/com/qingmei2/library/base/BaseApplication.java
package com.qingmei2.library.base;

import android.app.Activity;
import android.app.Application;
import android.support.v4.content.ContextCompat;

import com.qingmei2.library.di.module.AppModule;
import com.qingmei2.library.di.module.CacheModule;
import com.qingmei2.library.di.module.GlobalConfigModule;
import com.qingmei2.library.di.module.HttpClientModule;
import com.qingmei2.library.di.module.ServiceModule;
import com.qingmei2.library.http.Api;
import com.qingmei2.library.http.base.interceptor.HttpRequestHandler;

import javax.inject.Inject;

import dagger.android.AndroidInjector;
import dagger.android.DispatchingAndroidInjector;
import dagger.android.HasActivityInjector;
import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.logging.HttpLoggingInterceptor;

public class BaseApplication extends Application implements HasActivityInjector{
    @Inject
    DispatchingAndroidInjector<Activity> dispatchingAndroidInjector;

    private static BaseApplication instance;

    @Override
    public void onCreate() {
        super.onCreate();
        inject();
    }

    private void inject() {
        BaseApplication.instance = this;
        DaggerAppComponent.builder()
                .cacheModule(getCacheModule())
                .appModule(getAppModule())                      //注入application
                .globalConfigModule(getGlobalConfigModule())    //注入全局配置
                .httpClientModule(getHttpClientModule())        //注入http配置
                .serviceModule(getServiceModule())              //注入所有Service
                .build()
                .inject(this);
    }

    private AppModule getAppModule() {
        return new AppModule(this);
    }

    private GlobalConfigModule getGlobalConfigModule() {
        return GlobalConfigModule.buidler()
                .baseurl(Api.BASE_API)
                //这行代码为log打印网络请求信息，可以考虑在release版中取消该行代码
                .addInterceptor(new HttpLoggingInterceptor().setLevel(HttpLoggingInterceptor.Level.BODY))
                .globeHttpHandler(new HttpRequestHandler() {
                    @Override
                    public Response onHttpResultResponse(String httpResult, Interceptor.Chain chain, Response response) {
                        return response;
                    }

                    @Override
                    public Request onHttpRequestBefore(Interceptor.Chain chain, Request request) {
                        return request;
                    }
                })
                .build();
    }

    private HttpClientModule getHttpClientModule() {
        return new HttpClientModule();
    }

    private ServiceModule getServiceModule() {
        return new ServiceModule();
    }

    @Override
    public AndroidInjector<Activity> activityInjector() {
        return dispatchingAndroidInjector;
    }

    public static BaseApplication getInstance() {
        return instance;
    }

    public CacheModule getCacheModule() {
        return new CacheModule(ContextCompat.getExternalCacheDirs(this)[0]);
    }
}

File: rhine/src/main/java/com/qingmei2/rhine/base/BaseApplication.kt
package com.qingmei2.rhine.base

import android.app.Application
import android.content.Context
import com.qingmei2.rhine.di.HttpClientModule
import com.qingmei2.rhine.di.cacheModule
import com.qingmei2.rhine.di.rxModule
import com.qingmei2.rhine.di.serviceModule
import org.kodein.di.Kodein
import org.kodein.di.KodeinAware
import org.kodein.di.android.androidModule
import org.kodein.di.generic.bind
import org.kodein.di.generic.singleton

open class BaseApplication : Application(), KodeinAware {

    override val kodein: Kodein = Kodein.lazy {
        bind<Context>() with singleton { this@BaseApplication }
        import(androidModule(this@BaseApplication))

        import(serviceModule)
        import(cacheModule)
        import(rxModule)
        import(HttpClientModule)
    }

    override fun onCreate() {
        super.onCreate()
        INSTANCE = this
    }

    companion object {
        lateinit var INSTANCE: BaseApplication
    }
}
